README
NAME
    Class::Tiny::Antlers - Moose-like sugar for Class::Tiny

SYNOPSIS
       {
          package Point;
          use Class::Tiny;
          use Class::Tiny::Antlers;
          has x => (is => 'ro');
          has y => (is => 'ro');
       }
   
       {
          package Point3D;
          use Class::Tiny;
          use Class::Tiny::Antlers;
          extends 'Point';
          has z => (is => 'ro');
       }

DESCRIPTION
    Class::Tiny::Antlers provides Moose-like `has`, `extends` and `with`
    keywords for Class::Tiny. (The `with` keyword is implemented by
    Role::Tiny.)

    Class::Tiny doesn't support all Moose's attribute options; `has` should
    throw you an error if you try to do something it doesn't support (like
    triggers or type constraints).

    Class::Tiny::Antlers does however hack in support for `is => 'ro'` and
    Moo-style `is => 'rwp'`, clearers and predicates.

  Export
    By default, Class::Tiny::Antlers exports `has`, `with` and `extends`, and
    also imports strict into its caller. You can optionally also import
    `confess` and warnings:

       use Class::Tiny::Antlers qw( -default confess warnings );
       use Class::Tiny::Antlers qw( -all );   # same thing

    You can put a `no Class::Tiny::Antlers` statement at the end of your class
    definition to wipe the imported functions out of your namespace. (This
    does not unimport strict/warnings though.) To clean up your namespace more
    thoroughly, use something like namespace::sweep.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Class-Tiny-Antlers>.

SEE ALSO
    Class::Tiny, Role::Tiny.

    Moose, Mouse, Moo.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Class-Tiny-Antlers
==================

Created:      2013-08-29
Home page:    <https://metacpan.org/release/Class-Tiny-Antlers>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Class-Tiny-Antlers>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.022	2013-09-05

 [ Packaging ]
 - Add 'test requires' dependency on Test::Requires - oh, the irony!
 - Demote Role::Tiny and Class::Method::Modifiers from 'runtime recommends'
   dependencies to 'runtime suggests', but 'test recommends'.

0.021	2013-09-05

 - Actually inherit from Class::Tiny. This means that your class doesn't
   need to `use` both Class::Tiny and Class::Tiny::Antlers; just `use
   Class::Tiny::Antlers`.
 - Added: Optionally export `before`, `after` and `around` keywords,
   powered by an optional dependency on Class::Method::Modifiers.
 - Added: Provide a `get_all_attribute_specs_for` method for attribute
   introspection.
 - Make the dependency on Role::Tiny optional again.
 - Take advantage of Class::Tiny's new(ish) `create_attributes` method.
 - Updated: Requires Class::Tiny 0.006.

0.020	2013-08-30

 - Eliminate dependency on Package::Stash.

0.019	2013-08-30	Initial release; split from Moops 0.018


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = 'ExtUtils::MakeMaker'->VERSION;

my $meta = {
  "abstract"           => "Moose-like sugar for Class::Tiny",
  "author"             => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config"     => 0,
  "generated_by"       => "Dist::Inkt::Profile::TOBYINK version 0.007, CPAN::Meta::Converter version 2.120921",
  "keywords"           => [],
  "license"            => ["perl_5"],
  "meta-spec"          => {
                            url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                            version => 2,
                          },
  "name"               => "Class-Tiny-Antlers",
  "no_index"           => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"            => {
                            configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                            develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                            runtime   => {
                                           requires => { "Class::Tiny" => 0.006 },
                                           suggests => { "Class::Method::Modifiers" => 1.05, "Role::Tiny" => "1.000000" },
                                         },
                            test      => {
                                           recommends => { "Class::Method::Modifiers" => 1.05, "Role::Tiny" => "1.000000" },
                                           requires   => { "Test::Fatal" => 0.003, "Test::More" => 0.96, "Test::Requires" => 0.06 },
                                         },
                          },
  "provides"           => {
                            "Class::Tiny::Antlers" => { file => "lib/Class/Tiny/Antlers.pm", version => 0.022 },
                          },
  "release_status"     => "stable",
  "resources"          => {
                            bugtracker   => {
                                              web => "http://rt.cpan.org/Dist/Display.html?Queue=Class-Tiny-Antlers",
                                            },
                            homepage     => "https://metacpan.org/release/Class-Tiny-Antlers",
                            license      => ["http://dev.perl.org/licenses/"],
                            repository   => {
                                              type => "git",
                                              web  => "https://github.com/tobyink/p5-class-tiny-antlers",
                                            },
                            X_identifier => "http://purl.org/NET/cpan-uri/dist/Class-Tiny-Antlers/project",
                          },
  "version"            => 0.022,
  "x_provides_scripts" => {},
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT           => $meta->{abstract},
	AUTHOR             => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME           => $meta->{name},
	VERSION            => $meta->{version},
	EXE_FILES          => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME               => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} => $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my $minperl = delete $WriteMakefileArgs{PREREQ_PM}{perl};
	exists($WriteMakefileArgs{$_}) && delete($WriteMakefileArgs{$_}{perl})
		for qw(BUILD_REQUIRES TEST_REQUIRES CONFIGURE_REQUIRES);
	if ($minperl and $EUMM >= 6.48)
	{
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl;
	}
	elsif ($minperl)
	{
		die "Need Perl >= $minperl" unless $] >= $minperl;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

